pkg/apis/bindings/v1alpha1/imagebinding_lifecycle.go

```diff
@@ -26,25 +31,71 @@ func (b *ImageBinding) GetBindingStatus() duck.BindableStatus {
 	return &b.Status
 }
 
-func (b *ImageBinding) Do(ctx context.Context, ps *v1.WithPod) {
-	images := GetImages(ctx)
-	for _, p := range b.Spec.Providers {
-		for i, c := range ps.Spec.Template.Spec.Containers {
-			if c.Name == p.ContainerName {
-				img, ok := images[p.ContainerName]
-				if !ok {
-					continue
-				}
-				ps.Spec.Template.Spec.Containers[i].Image = img
-				ps.Annotations[ImageBindingAnnotationKey] = "bound"
-			}
+type imageBindingAnnotationData struct {
+	Container string `json:"container"`
+	Provider  string `json:"provider"`
+}
+
+func (b *ImageBinding) annotationKey() string {
+	return ImageBindingAnnotationKeyPrefix + b.Name
+}
+
+func (b *ImageBinding) annotationVal() string {
+	gvr, _ := meta.UnsafeGuessKindToResource(b.Spec.Provider.GroupVersionKind())
```

Comment:
we've seen some very wrong suggestions from this in the past. It is indeed unsafe and a guess.

Can we use `b.Spec.Provider.GroupVersionKind().String()`instead? Or just the kind?

or just drop the provider field since it's not actually used.

Reply:
We already use it for a more critical purpose here: https://github.com/projectriff/bindings/blob/master/pkg/resolver/imageable_resolver.go#L45

I'm happy to do something else for the annotation metadata. Do you have a better suggestion for the resolver logic?